# Webinar 2 - Intro to NLP, Text Processing, Spam Classifier
In this lession, We will go through the following topics one by one.
 1. **Intro to NLP**
 2. **Fetching data from a webpage**
 3. **Text processing in NLP**
 4. **Bayes theorem**
 5. **Naive bayes theorem**
 5. **Spam Detection using naive bayes theorem**
 
<hr>

## 1. Intro to NLP

In general we can define a NLP pipeline as geting a raw text, process it, extract relevant features, and build Modes to accomplish various NLP task. We can say that it can be divided into 3 parts:
1. **Text processing**
2. **Feature extraction**
3. **Modeling**

<hr>

## 2. Fetching data from a webpage

Let's go to <a href="https://techcrunch.com">techcrunch.com</a> and fetch all of the news titles.

<br>

<img src="assets/techcrunch.png">    

<br>

In [ ]:
# Importing the libraries
import request
from bs4 import BeautifulSoup

In [ ]:
# Fetching a webpage
r = requests.get("https://techcrunch.com")
print(r.text)

In [ ]:
# Remove the HTML tags
soup = BeautifulSoup(r.text, "html5lib")
print(soup.get_text())

In [ ]:
# Getting the titles of each article
summaries = soup.find_all("h2", class_="post-block__title")
summaries

In [61]:
# Getting the text of first article
summaries[0].find("a").get_text().strip()

'Breedr raises £2M led by LocalGlobe for its livestock data and trading platform'

In [62]:
# Getting the text of all articles
articles = [i_article.find("a").get_text().strip() for i_article in summaries]
articles

['Breedr raises £2M led by LocalGlobe for its livestock data and trading platform',
 'Google removed 2.3B bad ads, banned ads on 1.5M apps + 28M pages, plans new Policy Manager this year',
 'Google is reportedly shutting down its in-house VR film studio',
 'MIT’s deflated balloon robot hand can pick up objects 100x its own weight',
 'Tiny claws let drones perch like birds and bats',
 'Facebook won’t store data in countries with human rights violations — except Singapore',
 'The “splinternet” is already here',
 'Uber reportedly raising $1B in deal that values self-driving car unit at up to $10B',
 'Ford is expanding its self-driving vehicle program to Austin',
 'Ysplit wants to make it so you never owe your friends money again',
 'Inside Tufts University’s grade-hacking case',
 'DoorDash claims drivers made an average of at least $17.50/hour on deliveries in 2018',
 'Meet the 19 startups in AngelPad’s 12th batch',
 'Apple Music comes to Fire TV',
 'Take NVIDIA’s new Deep Learning Roboti

**Further Resources:**
1. <a href="https://www.youtube.com/watch?v=aIPqt-OdmS0">Web scraping and parsing with Beautiful Soup & Python Introduction p.1</a>
2. <a href="https://medium.freecodecamp.org/how-to-scrape-websites-with-python-and-beautifulsoup-5946935d93fe">How to scrape websites with Python and BeautifulSoup</a>
3. <a href="https://www.kdnuggets.com/2018/03/text-data-preprocessing-walkthrough-python.html">kdnuggets - Text Data Preprocessing: A Walkthrough in Python</a>

<hr>

## 3. Text Processing

In text processing we get the raw data and make it ready for feature extraction. The steps are as follows:
1. **Clean the dataset**: Here we will delete all the html tags, if there is some.
2. **Normalize the text**: First lowercase the dataset. Second remove punctuation. Third we tokenize the text.
3. **Remove stop words**
4. **Stemming**: Process of removing the suffixes. For example, “branching”, “branches”, “branched”can all be reduced to “branch”.
5. **Lemmatization**: This a technique for reducing words to a normalised form. In here we use a dictionary for mapping different variants of words back to its root. For example, converting words like “was”, “is”, “were” to “be”.

Stemming vs lemmatization: The final form of stemming can be meaningless (totally different word) but in lemmatization the final form is also a meaningful word in English. Also stemming doesn’t have a dictionary and its based on some rules but in lemmatization we have dictionary. So stemming less memory intensive option.

In [81]:
# Importing the libraries
import re
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from nltk.stem.wordnet import WordNetLemmatizer

### 3.1. Load and cleaning the dataset

In [82]:
text = "The first time you see The Second Renaissance it may look boring. Look at it at least twice and \
        definitely watch part 2. It will change your view of the matrix. Are the human people the ones  \
        who started the war ? Is AI a bad thing ?"

print(text)

The first time you see The Second Renaissance it may look boring. Look at it at least twice and         definitely watch part 2. It will change your view of the matrix. Are the human people the ones          who started the war ? Is AI a bad thing ?


### 3.2. Normalize the text

In [83]:
# Lower casing the dataset
text = text.lower()
print(text)

the first time you see the second renaissance it may look boring. look at it at least twice and         definitely watch part 2. it will change your view of the matrix. are the human people the ones          who started the war ? is ai a bad thing ?


In [84]:
# Remove punctuation
text = re.sub(r"[^a-zA-Z0-9]", " ", text)
print(text)

the first time you see the second renaissance it may look boring  look at it at least twice and         definitely watch part 2  it will change your view of the matrix  are the human people the ones          who started the war   is ai a bad thing  


In [85]:
# Tokenizing the text
text = word_tokenize(text)
print(text)

['the', 'first', 'time', 'you', 'see', 'the', 'second', 'renaissance', 'it', 'may', 'look', 'boring', 'look', 'at', 'it', 'at', 'least', 'twice', 'and', 'definitely', 'watch', 'part', '2', 'it', 'will', 'change', 'your', 'view', 'of', 'the', 'matrix', 'are', 'the', 'human', 'people', 'the', 'ones', 'who', 'started', 'the', 'war', 'is', 'ai', 'a', 'bad', 'thing']


### 3.3. Remove stop words

In [86]:
text = [i_word for i_word in text if i_word not in stopwords.words("english")]
print(text)

['first', 'time', 'see', 'second', 'renaissance', 'may', 'look', 'boring', 'look', 'least', 'twice', 'definitely', 'watch', 'part', '2', 'change', 'view', 'matrix', 'human', 'people', 'ones', 'started', 'war', 'ai', 'bad', 'thing']


### 3.4. Stemming

In [87]:
stemmed = [PorterStemmer().stem(i_word) for i_word in text]
print(stemmed)

['first', 'time', 'see', 'second', 'renaiss', 'may', 'look', 'bore', 'look', 'least', 'twice', 'definit', 'watch', 'part', '2', 'chang', 'view', 'matrix', 'human', 'peopl', 'one', 'start', 'war', 'ai', 'bad', 'thing']


### 3.5. Lemmatization

In [88]:
lemmatized = [WordNetLemmatizer().lemmatize(i_word, pos='n') for i_word in text]
print(lemmatized)

['first', 'time', 'see', 'second', 'renaissance', 'may', 'look', 'boring', 'look', 'least', 'twice', 'definitely', 'watch', 'part', '2', 'change', 'view', 'matrix', 'human', 'people', 'one', 'started', 'war', 'ai', 'bad', 'thing']


In [89]:
lemmatized = [WordNetLemmatizer().lemmatize(i_word, pos='v') for i_word in lemmatized]
print(lemmatized)

['first', 'time', 'see', 'second', 'renaissance', 'may', 'look', 'bore', 'look', 'least', 'twice', 'definitely', 'watch', 'part', '2', 'change', 'view', 'matrix', 'human', 'people', 'one', 'start', 'war', 'ai', 'bad', 'thing']


**Resources:**

1. <a href="https://web.stanford.edu/~jurafsky/slp3/ed3book.pdf">An Introduction to Natural Language Processing, Computational Linguistics, and Speech Recognition</a> - Chapter 2 Regular Expressions, Text Normalization, Edit Distance
2. <a href="https://www.kdnuggets.com/2018/03/text-data-preprocessing-walkthrough-python.html">kdnuggets - Text Data Preprocessing: A Walkthrough in Python</a>
3. <a href="https://pythonprogramming.net/tokenizing-words-sentences-nltk-tutorial/"> Natural Language Processing Series - Harrison</a>
4. <a href="https://medium.com/@datamonsters/text-preprocessing-in-python-steps-tools-and-examples-bf025f872908"> Text Preprocessing in Python: Steps, Tools, and Examples </a>

<hr>

## 4. Bayes Theorem

The **Bayes theorem** calculates the probability of an event occurring, based on certain other probabilities that are related to the event in question. It is  composed of a  prior(the probabilities that we are aware of or that is given to us) and the posterior(the probabilities we are looking to compute using the priors). 

##### Let's take a look at example for having a cancer:

Now let's find the odds of an individual having cancer, given that he or she was tested for it and got a positive result. 
Afterward we go on internet and search more about cancer and we figure out that 1% of whole population have cancer. we also figure out that the test correctly predicts 90% of times on those who have cancer (Sensitivity or True Positive Rate.) and it also correctly predicts 90% of times on those who doesn't have cancer. (Specificity or True Negative Rate.)

<br>

So we assume the following:

- `P(Cancer)`: The probability of a person having cancer. It's value is `0.01`  because 1% of whole population have cancer. 
<br>

- `P(Positive)`: The probability of getting a positive test result.
<br>

- `P(Negative)`: The probability of getting a negative test result.
<br>

- `P(Positve|Cancer)`: The probability of getting a positive result on a test done for detecting cancer, given that you have cancer. This has a value `0.9` because our test correctly predicts 90% of times on those who have a cancer.
<br>

- `P(Negative|~Cancer)`: The probability of getting a negative result on a test done for detecting cancer, given that you do not have cancer. This also has a value of `0.9` because our test correctly predicts 90% of times on those who doens't have a cancer.


The **Bayes theorem** formula is as follows:

<img src="assets/Bayes_theorem.png">

<br><br>

The probability of getting a positive test result `P(Positive)` can be calulated using the Sensitivity and Specificity:

<img src="assets/bayes_2.png">

In [99]:
### Now let's calculating P(Positive)

# P(Cancer)
p_cancer = 0.01

# P(~Cancer)
p_no_cancer = 1-0.01

# P(Positive|Cancer) or Sensitivity
p_positive_cancer = 0.9

# P(Negative|~Cancer) or Specieficity
p_negative_no_cancer = 0.9

# P(Positive)
p_positive = (p_cancer * p_positive_cancer) + (p_no_cancer * (1-p_negative_no_cancer))
print("P(Positve) or probability of getting a positive test result is equal to: {:.2f}%".format(p_positive*100))

P(Positve) or probability of getting a positive test result is equal to: 10.80%


<br><br>
Given all of this information we can calculate our **posteriors** as follows:
<img src="assets/posterior_1.png">
<img src="assets/posterior_2.png">

In [100]:
### Calculating P(Cancer|Positive)

p_cancer_positive = (p_positive_cancer * p_cancer) / p_positive
print("P(Cancer|Positive) or Probability of having cancer given that the test is positive: {:.2f}%".format(p_cancer_positive*100))

P(Cancer|Positive) or Probability of having cancer given that the test is positive: 8.33%


In [104]:
### Calculating P(~Cancer|Positive)

p_pos_no_cancer = 1 - p_negative_no_cancer

p_no_cancer_positive = (p_no_cancer * p_pos_no_cancer) / p_positive

print("P(~Cancer|Positive) or Probability of not having cancer given that the test is positive: {:.2f}%".format(p_no_cancer_positive*100))

P(~Cancer|Positive) or Probability of not having cancer given that the test is positive: 91.67%


Note that since: `P(~Cancer|Positive) = (P(~Cancer) * P(Positive|~Cancer) / P(Positive)`

Then `P(Positive/~Cancer)` can be computed as `1 - P(Negative/~Cancer)`. 

##### Our result shows that if we get a positive test result, then there is only a 8.33% chance that we actually have cancer and a 91.67% chance that we do not have cancer.

**Further Resources:**

1. <a href="https://classroom.udacity.com/courses/st101">intro to statistics (Bayes rule)</a> - Udacity free course
2. <a href="https://betterexplained.com/articles/an-intuitive-and-short-explanation-of-bayes-theorem/">An Intuitive (and Short) Explanation of Bayes’ Theorem</a> 
3. <a href="https://www.mathsisfun.com/data/bayes-theorem.html">Bayes can do magic!</a> 

<hr>

## 5. Naive Bayes Theorem

The word naive in naive bayes come from the naive assumption we are doing. we always assume that the events are independent from each other. This is a false and naive assumption but in practice it words pretty well.

<img src="assets/naive assumption.png">

##### Let's take an example in political usage:
Assume we have two political candidates in two different parties, 'Jennifer Stewart' of the Gemini Party and 'Gabrielle Jones' of the Liberal Party. Assume that the probablility of Jennifer Stewart giving a speech, `P(Jenifer)` is `0.5` and the same for Gabrielle Jones, `P(Gabrielle) = 0.5`. Below we have the probabilities of each of these candidates saying the words 'liberty', 'war' and 'economy' when they give a speech:

- `P(Liberty|Jenifer)` or Probability saying 'liberty' given Jennifer Stewart says it: `0.1`
<br>
- `P(War|Jenifer)` or Probability saying 'war' given Jennifer Stewart says it: `0.1`
<br>
- `P(Economy|Jenifer)` or Probability saying 'economy' given Jennifer Stewart says it: `0.8`
<br><br>
- `P(Liberty|Gabrielle)` or Probability saying 'liberty' given Gabrielle Jones says it: `0.7`
<br>
- `P(War|Gabrielle)` or Probability saying 'war' given Gabrielle Jones says it: `0.2`
<br>
- `P(Economy|Gabrielle)` or Probability saying 'economy' given Gabrielle Jones says it: `0.1`
<br>



**Naive assumption:** Given this, what if we had to find the probabilities of Jennifer Stewart saying the words 'liberty' and 'war'? This is where the Naive Bayes'theorem comes into play as we are considering two features, 'liberty' and 'war'.

<br><br>
Now we are at a place where we can define the <u>formula for the Naive Bayes' theorem</u>:

<img src="assets/naivebayes.png" height="342" width="342">

- `y` is the class variable or in our case the name of the candidate 
- `x1` through `xn` are the feature vectors or in our case the individual words. The theorem makes the assumption that each of the feature vectors or words (`xi`) are independent of each other.
<br><br>

**To break this down, we have to compute the following posterior probabilities:**

1. `P(Jenifer|Liberty,War)`: Probability of Jennifer Stewart saying the words liberty and war. Using the formula and our knowledge of Bayes' theorem, we can compute this as follows:
    
    <img src="assets/jenifer_naivebayes.png">
    
    Here P(Liberty,War) is the probability of the words 'liberty' and 'war' being said in a speech.
    

2. `P(Gabrielle|Liberty,War)`: Probability of Gabrielle Jones saying the words liberty and war. Using the formula, we can compute this as follows:
    
    <img src="assets/gabrielle_naivebayes.png">
    
    

##### Now compute the `P(Liberty,War)` or the probability of the words 'liberty' and 'war' being said in a speech

In [116]:
### Computing the 3 different probability for Jenifer

# P(Jenifer)
p_jenifer = 0.5

# P(Libery | Jenifer)
p_liberty_jenifer = 0.1

# P(War | Jenifer)
p_war_jenifer = 0.1

In [117]:
### Computing the 3 different probability for Gabrielle

# P(Gabrielle)
p_gabrielle = 0.5

# P(Liberty | Gabrielle)
p_liberty_gabrielle = 0.7

# P(War | Gabrielle)
p_war_gabrielle = 0.2

In [118]:
### Compute P(Liberty , War)

# [P(Jenifer) x P(Libery|Jenifer) P(War|Jenifer)] + [P(Gabrielle) x P(Liberty|Gabrielle) x P(War|Gabrielle)]
p_liberty_war = (p_jenifer * p_liberty_jenifer * p_war_jenifer) + (p_gabrielle * p_liberty_gabrielle * p_war_gabrielle)

print("Prbability of the words 'liberty' and 'war' being said in a speech: {:.2f}%".format(p_liberty_war*100))

Prbability of the words 'liberty' and 'war' being said in a speech: 7.50%


In [119]:
### Compute P(Jenifer | Liberty, War) or the posterior probability of Jenifer

# P(Jenifer|Libery,War) = (P(Jenifer) * P(Libery|Jenifer) * P(War|Jenifer)) / P(Libery,War)
p_jenifer_liberty_war = (p_jenifer * p_liberty_jenifer * p_war_jenifer) / p_liberty_war

print("The probability the Jenifer saying the word liberty and war is {:.2f}%".format(p_jenifer_liberty_war*100))

The probability the Jenifer saying the word liberty and war is 6.67%


In [120]:
### Compute P(Gabrielle | Liberty, War) or the posterior probability of Gabrielle

# P(Gabrielle|Libery,War) = (P(Gabrielle) * P(Libery|Gabrielle) * P(War|Gabrielle)) / P(Libery,War)
p_gabrielle_liberty_war = (p_gabrielle * p_liberty_gabrielle * p_war_gabrielle) / p_liberty_war

print("The probability the Gabrielle saying the word liberty and war is {:.2f}%".format(p_gabrielle_liberty_war*100))

The probability the Gabrielle saying the word liberty and war is 93.33%


**Note: The sum of our posterior probability should add up to 1**

**Further Resources:**

1. <a href="https://classroom.udacity.com/courses/ud120">intro to machine learning (Naive Bayes)</a> - Udacity free course
2. <a src="https://www.analyticsvidhya.com/blog/2017/09/naive-bayes-explained/">6 Easy Steps to Learn Naive Bayes Algorithm (with codes in Python and R)</a>
3. <a src="https://towardsdatascience.com/naive-bayes-classifier-81d512f50a7c">Naive Bayes Classifier</a>

<hr>

## 6. Spam Detection Using Naive Bayes Theorem

### Introduction 

Naive Bayes classifiers are a popular statistical technique of e-mail filtering. They typically use bag of words features to identify spam e-mail, an approach commonly used in text classification.

Naive Bayes classifiers work by correlating the use of tokens (typically words, or sometimes other things), with spam and non-spam e-mails and then using Bayes' theorem to calculate a probability that an email is or is not spam.

Naive Bayes spam filtering is a baseline technique for dealing with spam that can tailor itself to the email needs of individual users and give low false positive spam detection rates that are generally acceptable to users. It is one of the oldest ways of doing spam filtering, with roots in the 1990s. (Wikipedia)

In [1]:
# Import the libraries
import pandas as pd
import re
from collections import Counter
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

### Loading the Dataset
We will be using the Naive Bayes algorithm to create a model that can classify [dataset](https://archive.ics.uci.edu/ml/datasets/SMS+Spam+Collection) SMS messages as spam or not spam, based on the training we give to the model. 

In [3]:
# Loading the dataset
dataset = pd.read_table("./dataset/SMSSpamCollection", 
                        sep = "\t", 
                        header = None, 
                        names = ["label", "sms_message"])
dataset.head(10)

label                                        sms_message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...
5  spam  FreeMsg Hey there darling it's been 3 week's n...
6   ham  Even my brother is not like to speak with me. ...
7   ham  As per your request 'Melle Melle (Oru Minnamin...
8  spam  WINNER!! As a valued network customer you have...
9  spam  Had your mobile 11 months or more? U R entitle...

In [4]:
print("dataset shape: ", dataset.shape)

dataset shape:  (5572, 2)


### Data Preprocessing

Now that we loaded our dataset, Let's convert our labels into a binary number. 0 for ham and 1 for spam.

In [5]:
# Converting the labels to binary numbers
dataset["label"] = dataset["label"].map({"ham": 0, "spam": 1})
dataset.head(10)

label                                        sms_message
0      0  Go until jurong point, crazy.. Available only ...
1      0                      Ok lar... Joking wif u oni...
2      1  Free entry in 2 a wkly comp to win FA Cup fina...
3      0  U dun say so early hor... U c already then say...
4      0  Nah I don't think he goes to usf, he lives aro...
5      1  FreeMsg Hey there darling it's been 3 week's n...
6      0  Even my brother is not like to speak with me. ...
7      0  As per your request 'Melle Melle (Oru Minnamin...
8      1  WINNER!! As a valued network customer you have...
9      1  Had your mobile 11 months or more? U R entitle...

### Splitting the Dataset
In order to find a model with the highest accuracy possible, We need further testing for our model. For doing so we split our dataset into training and testing set. One for training and another one for testing the model.

In [6]:
# Splitting the dataset into training and test set
X_train, X_test, y_train, y_test = train_test_split(dataset["sms_message"],
                                                    dataset["label"],
                                                    test_size =0.25,
                                                    random_state = 1)

print("Training set size: \n", X_train.shape[0])
print("Test set size: \n", X_test.shape[0])

Training set size: 
 4179
Test set size: 
 1393


### Feature Extraction - Bag of Words
Now that we have split the data, our next objective is to get the Bag of words and convert our data into the desired matrix format. To do this we will be using CountVectorizer(). The steps are as follows:
* First we fir the `CountVectorizer()` into the training data or `X_train`.
* Then we transform our testing data or `X_test`

In [7]:
# Initializing the count vectorizer
count_vector = CountVectorizer(lowercase = True, 
                               token_pattern = "(?u)\\b\\w\\w+\\b",
                               stop_words = "english")

# Fit and transform the training data
training_data = count_vector.fit_transform(X_train)
#print("Vocabulary in Training Set: \n", count_vector.get_feature_names(), "\n\n")

# Transform the test data
test_data = count_vector.transform(X_test)

In [8]:
# Checking the training data in DataFrame
pd.DataFrame(data = training_data.toarray(), 
             columns = count_vector.get_feature_names()).head(10)

00  000  008704050406  0121  01223585236  01223585334  0125698789  02  \
0   0    0             0     0            0            0           0   0   
1   0    0             0     0            0            0           0   0   
2   0    0             0     0            0            0           0   0   
3   0    0             0     0            0            1           0   0   
4   0    0             0     0            0            0           0   0   
5   0    0             0     0            0            0           0   0   
6   0    0             0     0            0            0           0   0   
7   0    0             0     0            0            0           0   0   
8   0    0             0     0            0            0           0   0   
9   0    0             0     0            0            0           0   0   

   0207  02072069400 ...   zed  zeros  zhong  zindgi  zoe  zoom  zouk  zyada  \
0     0            0 ...     0      0      0       0    0     0     0      0   
1     0            0 ...     0      0      0       0    0     0     0      0   
2     0            0 ...     0      0      0       0    0     0     0      0   
3     0            0 ...     0      0      0       0    0     0     0      0   
4     0            0 ...     0      0      0       0    0     0     0      0   
5     0            0 ...     0      0      0       0    0     0     0      0   
6     0            0 ...     0      0      0       0    0     0     0      0   
7     0            0 ...     0      0      0       0    0     0     0      0   
8     0            0 ...     0      0      0       0    0     0     0      0   
9     0            0 ...     0      0      0       0    0     0     0      0   

   èn  〨ud  
0   0    0  
1   0    0  
2   0    0  
3   0    0  
4   0    0  
5   0    0  
6   0    0  
7   0    0  
8   0    0  
9   0    0  

[10 rows x 7204 columns]

### Naive Bayes

Sklearn has several Naive Bayes implementations that we can use and so we do not have to do the math from scratch. We will be using sklearns `sklearn.naive_bayes` method.

We will be using the multinomial Naive Bayes implementation. This particular classifier is suitable for classification with discrete features.

In [9]:
# Applying Naive Bayes
naive_bayes = MultinomialNB()

naive_bayes.fit(training_data, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

### Prediction
Let's make a prediction on our testing dataset. Later on we will evaluate how well did our model do on this dataset.

In [10]:
# Predicting the test set
predictions = naive_bayes.predict(test_data)
print("predictions: ", predictions)

predictions:  [0 0 0 ... 0 1 0]


In [11]:
# Converting the prediction "0", "1" into "spam", "ham"
def pred_num_to_text(prediction):
    if prediction == 1:
        return "spam"
    elif prediction == 0:
        return "ham"

In [12]:
id_msg = 16

print("The recieved message: \n\"" + X_test.iloc[id_msg] + "\"\n")
print("Prediction: \n", pred_num_to_text(naive_bayes.predict(test_data[id_msg])))

The recieved message: 
"FreeMsg Why haven't you replied to my text? I'm Randy, sexy, female and live local. Luv to hear from u. Netcollex Ltd 08700621170150p per msg reply Stop to end"

Prediction: 
 spam


In [13]:
id_msg = 6

print("The recieved message: \n\"" + X_test.iloc[id_msg] + "\"\n")
print("Prediction: \n", pred_num_to_text(naive_bayes.predict(test_data[id_msg])))

The recieved message: 
"I want kfc its Tuesday. Only buy 2 meals ONLY 2. No gravy. Only 2 Mark. 2!"

Prediction: 
 ham


### Evaluating the Model

Now we want to evaluate how well our model is doing. There are various mechanisms for doing so:

1. **Accuracy**: `[Correct Predictions/Total Number of Predictions]`

2. **Precision**: `[True Positives/(True Positives + False Positives)]`

3. **Recall(sensitivity)**: `[True Positives/(True Positives + False Negatives)]`

4. **F1 score**: weighted average of the precision and recall scores. This score can range from 0 to 1, with 1 being the best possible F1 score.

In [14]:
# Evaluation
print("Accuracy: {:.2f}%".format(accuracy_score(y_test, predictions)*100))
print("Precision Score: {:.2f}%".format(precision_score(y_test, predictions)*100))
print("Recall Score: {:.2f}%".format(recall_score(y_test, predictions)*100))
print("F1 Score: {:.2f}%".format(f1_score(y_test, predictions)*100))

Accuracy: 98.78%
Precision Score: 96.15%
Recall Score: 94.59%
F1 Score: 95.37%


**Resources:**

<a src="http://www.est.uc3m.es/BayesUC3M/Summer_School_UPM/2017/lecture%20notes/Practical1.pdf">Case Study I: Naive Bayesian spam filtering</a> - Universidad Carlos III de Madrid